File: Backend/Services/UserService/OnlineExamSystem.User.Domain/Entities/Plan.cs
using OnlineExamSystem.Shared.Common.Entities;
using OnlineExamSystem.Shared.Common.Multitenancy;

namespace OnlineExamSystem.User.Domain.Entities;

// Not tenant-scoped itself - a Plan is a platform-wide template Super Admin
// defines once and assigns to any number of Tenants (Tenant.PlanId).
public class Plan : BaseEntity
{
    public required string Name { get; set; }
    public string? Description { get; set; }
    public List<PlanFeature> IncludedFeatures { get; set; } = [];

    // Stored + displayed reference values only - no billing/payment system
    // exists anywhere in this codebase, and this deliberately doesn't add
    // one. Never used in any calculation.
    public decimal? MonthlyPrice { get; set; }
    public decimal? AnnualPrice { get; set; }

    // The entitlement source for Tenant's own MaxUsers/MaxAdmins/
    // MaxInstructors/MaxStudents/MaxExams ("effective limits" - seeded from
    // here at tenant creation and on every real plan change, see
    // CreateTenantHandler/AssignPlanToTenantHandler). Nullable = unlimited,
    // same convention Tenant already used before this existed.
    public int? MaxStudents { get; set; }
    public int? MaxAdmins { get; set; }
    public int? MaxInstructors { get; set; }
    public int? MaxExams { get; set; }

    // Persisted + displayed only - deliberately NOT enforced anywhere.
    // No question-count check, no monthly AI-generation usage counter, and
    // no storage-accounting system exist anywhere in this codebase; adding
    // real enforcement for these would mean building that metering
    // infrastructure from scratch, which is explicitly out of scope here.
    public int? MaxQuestions { get; set; }
    public int? MaxAiQuestionsPerMonth { get; set; }
    public int? StorageGb { get; set; }

    public DateTime UpdatedAtUtc { get; set; } = DateTime.UtcNow;

    // Which Super Admin created/last edited this pricing plan - real
    // accountability for a platform-wide entitlement template, not
    // previously tracked. Nullable only because pre-existing seeded plans
    // predate this field.
    public Guid? CreatedByUserId { get; set; }
    public Guid? UpdatedByUserId { get; set; }
}

File: Frontend/examvault-web/src/types/plan.ts
// Mirrors the backend's PlanFeature enum (Backend/Shared/.../Multitenancy/PlanFeature.cs)
// - the 10 gateable Admin console modules. ExamSecurity/Proctoring split
// out of LiveMonitoring 2026-09-04 (see ActionPlan.txt's "SPLIT
// LiveMonitoring" plan) - LiveMonitoring narrowed to non-video real-time
// oversight, ExamSecurity covers violation tracking, Proctoring covers
// webcam/live-video/recording.
export type PlanFeature =
  | 'Users'
  | 'Exams'
  | 'ExamTypes'
  | 'LiveMonitoring'
  | 'Results'
  | 'Reports'
  | 'Notifications'
  | 'Settings'
  | 'ExamSecurity'
  | 'Proctoring';

export const ALL_PLAN_FEATURES: PlanFeature[] = [
  'Users',
  'Exams',
  'ExamTypes',
  'LiveMonitoring',
  'Results',
  'Reports',
  'Notifications',
  'Settings',
  'ExamSecurity',
  'Proctoring',
];

export const PLAN_FEATURE_LABELS: Record<PlanFeature, string> = {
  Users: 'User Management',
  Exams: 'Exam Authoring',
  ExamTypes: 'Exam Types',
  LiveMonitoring: 'Live Monitoring',
  Results: 'Results',
  Reports: 'Reports',
  Notifications: 'Notifications',
  Settings: 'Settings',
  ExamSecurity: 'Exam Security',
  Proctoring: 'Proctoring',
};

// Pricing + usage-limit fields added 2026-09-04 (ActionPlan.txt's "PLAN
// PRICING & PER-PLAN USAGE LIMITS" plan). MonthlyPrice/AnnualPrice are
// stored/displayed only - no billing provider exists yet. Of the 7 limit
// fields, only maxStudents/maxAdmins/maxInstructors/maxExams are actually
// enforced server-side (Tenant's effective limits, seeded from these on
// tenant creation/plan change); maxQuestions/maxAiQuestionsPerMonth/storageGb
// are persisted and shown but NOT enforced - their metering infrastructure
// doesn't exist yet.
export interface Plan {
  id: string;
  name: string;
  description: string | null;
  includedFeatures: PlanFeature[];
  monthlyPrice: number | null;
  annualPrice: number | null;
  maxStudents: number | null;
  maxAdmins: number | null;
  maxInstructors: number | null;
  maxExams: number | null;
  maxQuestions: number | null;
  maxAiQuestionsPerMonth: number | null;
  storageGb: number | null;
  createdAtUtc: string;
  updatedAtUtc: string;
}

export interface CreatePlanRequest {
  name: string;
  description: string | null;
  includedFeatures: PlanFeature[];
  monthlyPrice: number | null;
  annualPrice: number | null;
  maxStudents: number | null;
  maxAdmins: number | null;
  maxInstructors: number | null;
  maxExams: number | null;
  maxQuestions: number | null;
  maxAiQuestionsPerMonth: number | null;
  storageGb: number | null;
}

export interface UpdatePlanRequest {
  name: string;
  description: string | null;
  includedFeatures: PlanFeature[];
  monthlyPrice: number | null;
  annualPrice: number | null;
  maxStudents: number | null;
  maxAdmins: number | null;
  maxInstructors: number | null;
  maxExams: number | null;
  maxQuestions: number | null;
  maxAiQuestionsPerMonth: number | null;
  storageGb: number | null;
}
